bus-mapping/src/evm/memory.rs

```diff
@@ -22,30 +19,59 @@ impl MemoryAddress {
 
     /// Return the little-endian byte representation of the word as a 32-byte
     /// array.
-    pub fn to_bytes(self) -> [u8; 32] {
+    pub fn to_le_bytes(self) -> [u8; 32] {
         let mut array = [0u8; 32];
         let bytes = self.0.to_le_bytes();
         array[..bytes.len()].copy_from_slice(&bytes[0..bytes.len()]);
 
         array
     }
 
-    /// Generate a MemoryAddress from the provided set of bytes.
-    pub fn from_bytes<T: AsRef<[u8]>>(bytes: T) -> Self {
+    /// Return the big-endian byte representation of the word as a 32-byte
+    /// array.
+    pub fn to_be_bytes(self) -> [u8; 32] {
+        let mut array = [0u8; 32];
+        let bytes = self.0.to_be_bytes();
+        array[32 - bytes.len()..].copy_from_slice(&bytes[0..bytes.len()]);
+
+        array
+    }
+
+    /// Generate a MemoryAddress from the provided set of little-endian bytes.
+    pub fn from_le_bytes<T: AsRef<[u8]>>(bytes: T) -> Result<Self, Error> {
+        if bytes.as_ref().len() > core::mem::size_of::<usize>() {
+            return Err(Error::InvalidMemoryPointer);
+        }
         let mut array = [0u8; core::mem::size_of::<usize>()];
         array[..]
             .copy_from_slice(&bytes.as_ref()[..core::mem::size_of::<usize>()]);
-        MemoryAddress::from(usize::from_le_bytes(array))
+        Ok(MemoryAddress::from(usize::from_le_bytes(array)))
+    }
+
+    /// Generate a MemoryAddress from the provided set of big-endian bytes.
+    pub fn from_be_bytes<T: AsRef<[u8]>>(bytes: T) -> Result<Self, Error> {
+        if bytes.as_ref().len() > core::mem::size_of::<usize>() {
+            return Err(Error::InvalidMemoryPointer);
+        }
+        let mut array = [0u8; core::mem::size_of::<usize>()];
+        array[..]
+            .copy_from_slice(&bytes.as_ref()[..core::mem::size_of::<usize>()]);
+        Ok(MemoryAddress::from(usize::from_be_bytes(array)))
     }
 }
 
 impl TryFrom<EvmWord> for MemoryAddress {
     type Error = Error;
 
     fn try_from(word: EvmWord) -> Result<Self, Self::Error> {
-        let addr: usize =
-            word.0.try_into().map_err(|_| Error::MemAddressParsing)?;
-        Ok(MemoryAddress(addr))
+        let (should_be_zeroes, usize_bytes) =
+            word.inner().split_at(32 - core::mem::size_of::<usize>());
+        if should_be_zeroes == &[0] {
```

Comment:
Why only leaving with 1 zero failes? I assume you are checking there should be no non-zero value left in first part.

BTW just try this but it succeeds.

```rust
assert_eq!(
    MemoryAddress::try_from(EvmWord::from(u64::MAX))?,
    MemoryAddress::from(usize::MAX)
);
```

Reply:
Ah you are right, my mistake, they are identical on my machine.